core/src/main/scala/kafka/controller/KafkaController.scala

```diff
@@ -284,12 +289,18 @@ class KafkaController(val config : KafkaConfig, zkUtils: ZkUtils, val brokerStat
                   }
                   // If the broker is a follower, updates the isr in ZK and notifies the current leader
                   replicaStateMachine.handleStateChanges(Set(PartitionAndReplica(topicAndPartition.topic,
-                    topicAndPartition.partition, id)), OfflineReplica)
+                    topicAndPartition.partition, id)), OfflineReplica, newBatch = false)
                 }
               }
             }
           }
       }
+
+      inLock(controllerContext.controllerLock) {
```

Comment:
@lindong28 can you double-check your test results below? We put in a similar hot-fix at LinkedIn to temporarily address long shutdowns in some of our larger clusters (with ~ 30k partitions). The 37% savings that you note (3.8 to 2.4) don't seem to line up with the improvements we saw with that patch.

As I suggested above, I'm not very comfortable with effectively removing the assertion especially if the savings are not significant.